### DISCUSSION TITLE
Documenting cmd2 project

### DISCUSSION THREAD

**Discussion by xuoguoto** [mkdocstrings/mkdocstrings#480] [Q&A] [answered]
Hello all,

I am working with an application which uses `cmd2` library. `cmd2` internally uses `argparse` for  parsing commands and it has defined some helpful [decorators](https://cmd2.readthedocs.io/en/stable/examples/first_app.html#create-a-command) for easy coding. `cmd2` has facility for displaying help text for all commands and options and its nicely integrated into the program. For example given a program like the following:

```python
#!/usr/bin/env python
"""A simple cmd2 application."""
import cmd2


class FirstApp(cmd2.Cmd):
    """A simple cmd2 application."""

    def __init__(self):
        super().__init__()

    speak_parser = cmd2.Cmd2ArgumentParser(description='This command repeats what you tell me to.')
    speak_parser.add_argument('-p', '--piglatin', action='store_true', help='atinLay')
    speak_parser.add_argument('-s', '--shout', action='store_true', help='N00B EMULATION MODE')
    speak_parser.add_argument('words', nargs='+', help='words to say')

    @cmd2.with_argparser(speak_parser)
    def do_speak(self, args):
        self.poutput(args.words)

if __name__ == '__main__':
    import sys
    c = FirstApp()
    sys.exit(c.cmdloop())
```

It can automatically show the following output:

```
$ python cmd2-sample.py
(Cmd) speak -h
Usage: speak [-h] [-p] [-s] words [...]

This command repeats what you tell me to.

positional arguments:
  words           words to say

optional arguments:
  -h, --help      show this help message and exit
  -p, --piglatin  atinLay
  -s, --shout     N00B EMULATION MODE

(Cmd)
```

Now for a commdline line application, its essential to provide a command reference and the best way to do it is extract it from the code.  If I can use mkdocstrings to export all the help into static web pages using mkdocs, that will make a nice command reference which gets updated as code is changed automatically. 

My problem now is that I am unable to figure out a means to pass this help text into docstrings or by some other means to mkdocstrings for processing.

Any help in figuring it out will be much appreciated. Thanks!

**Comment by pawamoy** [comment, marked as answer]
Hello @xuoguoto, I've converted your issue as a discussion :slightly_smiling_face: 

mkdocstrings is probably not best-suited for this use-case. mkdocstrings is meant for API docs, while you're looking for something more akin to CLI docs. You'd typically enjoy something like [mkdocs-click](https://pypi.org/project/mkdocs-click/) but for cmd2 :slightly_smiling_face: Maybe you could port it? Otherwise, an other solution would be to use [markdown-exec](https://pawamoy.github.io/markdown-exec/) to iterate on your CLI parser and build some Markdown/HTML. [PDM](https://github.com/pdm-project/pdm) documents its CLI using it: see [the code](https://github.com/pdm-project/pdm/blob/main/docs/docs/usage/cli_reference.md) (click on the "Raw" button to see that the code block is executed thanks to markdown-exec!) and [the rendered HTML page](https://pdm.fming.dev/latest/usage/cli_reference/).

Hope this helps!

**Comment by xuoguoto** [comment]
Thanks for the helpful hints. I will chase them and see where it goes :)
